README.md
cloudwatch-buddy
===============

WARNING: This is currently in "beta" and has not been extensively tested. Please wait for a 1.0.0 release before using in production. Until then, feel free to create issues or pull requests.

## Description

cloudwatch-buddy is a node module which easily allows for the sending of CloudWatch metrics, statistics, and logs to AWS CloudWatch. Using this module, you can easily replace StatsD with AWS CloudWatch. It gracefully hadles single increments (such as pageviews), as well as more complex measurements (page load times or size) with sum, minimums, and maximums with custom dimensions. Additionally, it can stream logs to AWS, dealing with the timestamps and formatting issues. It also manages periodic sending of logs and the AWS "next token" pattern, as well as uploading copies of the log files to an S3 bucket of your choosing.

## Features

* Simple "increment" measurements
* Statistics (sum, minimum, maximum) across a timeframe
* Custom dimensions
* Periodic uploading
* Log ordering
* Retrying of failed data submissions
* Sending 0 as value so application has data
* S3 bucket uploads

## Usage

```
npm install cloudwatch-buddy
```

```
var CloudWatchBuddy = require('cloudwatch-buddy');

var awsOptions = {
	accessKeyId: 'access',		// Optional. It is suggested to use IAM roles (see permissions below)
	secretAccessKey: 'secret',	// Optional
	sessionToken: 'session',	// Optional
	region: 'us-east-1'			// Required. You must enter a valid AWS region
};

var metricsOptions = {			// You can use either or both metric and log collection.
	namespace: 'test-data',
	timeout: 60					// See below for a complete list of options
};

var logsOptions = {
	logGroup: 'my-application',
	timeout: 60,
	maxSize: 10000,
	addInstanceId: true,
	addTimestamp: true,
	logFormat: 'string',
	debug: true,
	s3Bucket: 'mybucket.example.com',
	s3Prefix: 'app/test',
	s3Subfolders: true
};

var cwbMetrics = new CloudWatchBuddy(awsOptions).metrics(metricsOptions);
var cwbLogs = new CloudWatchBuddy(awsOptions).logs(logsOptions);

// Submit a simple Count metric for page views
cwbMetrics.increment('pageviews');

// Submit a metric with a unit, keeping track of sum, minimum, maximum
cwbMetrics.stat('loadtime', 10, 'Milliseconds');

// Submit a metric with a unit, with custom dimensions
cwbMetrics.stat('serverload', 10, 'Percent', {
	serverName:'web01.example.com',region:'us-east-1'
});

// Send a log
cwbLogs.log('errors', 'Test message');
cwbLogs.log('signups', 'New user');
```

## Options

### AWS Options

The AWS options object must be a valid config object supported by AWS (see: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html). You *must* provide a region.

### Metrics

The following metrics options are supported:

`namespace` (Required) - The AWS name space that all metrics should be submitted under. Note: to use multiple namespaces, you must instantiate separate metrics objects.

`timeout` - The interval, in seconds, to submit your metrics. The minimum is 60 seconds (AWS CloudWatch charts only show minute-level metrics) and the maximum is 30 minutes (1800 s). The default is 120 seconds.

`debug` - Whether messages should be printed to the console; helps with debugging

### Logs

The following logs options are supported:

`logGroup` (Required) - The log group name in AWS. This must already exist (the module will create streams for you, but not log groups).

`timeout` - The interval, in seconds, to submit your logs. The minimum is 60 seconds and the maximum is 30 minutes (1800 s). The default is 120 seconds.

`maxSize` - The maximum size, in bytes, of a log stream group between uploads. The minimum is 5000 and the maximum is 1048576 (~1MB). The default is 200000 (200KB). Note that this is a *rough estimate*.

`logFormat` - Either "string" or "json". For string, the logs are written as timestamp - instance id - log message. For JSON, it is saved as an object with these same keys. The default is "string".

`addTimestamp` - Whether the timestamp should be included in the log message. All logs have timestamps sent to AWS (used for ordering), but this option allows the timestamp to be included in the actual log message contents as well. The default is false.

`addInstanceId` - Whether the EC2 instance ID should be included in the log message. The default is false. The module will attempt to load the instance ID from the AWS metadata service. If it cannot be determined, the instance ID will become "unknown".

`debug` - Whether messages should be printed to the console; helps with debugging

`s3Bucket` - If included, copies of the logs will be uploaded to this bucket (also requires s3Prefix to be set).

`s3Prefix` - If included, copies of the logs will be uploaded to this prefix within the bucket (also requires s3Bucket to be set).

`s3Subfolders` - Whether logs should be put in subfolders by year / month / day. (If true, path would be bucket/2015/04/15/22-45-153.log vs bucket/2015-04-15-22-45-153.log).

## Permissions

You can either hard code AWS credentials into the application (please don't do this unless you're testing locally) or use IAM roles. When using IAM roles, your instance must have the following permissions in order to allow cloudwatch-buddy to create log streams, put log events, and put metrics:

```
{
	"Effect" : "Allow",
	"Action" : [
		"cloudwatch:PutMetricData",
		"logs:CreateLogStream",
		"logs:DescribeLogGroups",
		"logs:DescribeLogStreams",
		"logs:PutLogEvents"
	],
	"Resource" : "*"
}
```

## Known Limitations

* Because of the 40KB limit imposed by API calls to CloudWatch, the number of stats that can be recorded is limited to approximately 100. This means that you can have 100 separate metric names ("page load time," "page size," etc.). You can update them as often as needed, either as increments or with dimensions, but you cannot create too many. Note: this is probably a good limit, as AWS imposes limits on the number of distinct metric names you can have as well.

* The log group must exist on AWS before the module can write the logs. Future versions may allow for the log group to be created automatically, but it would require more permissions than are preferred. At this point, simply create the log group for your application before running this module.

RELEASES.md
### 0.0.14
* Fixed bug where no space was added between instance ID and the message for string-format logs
* Fixed bug causing successive log streams to be skipped if the first failed

### 0.0.13
* Fixed bug where unit was overwritten when multiple stats submitted in one group

### 0.0.12
* Fixed bug causing minimum metric value to be set to zero when only one sample

### 0.0.11
* If sample count is 0, change to 1 but keep values as 0 so the charts maintain a connected line

### 0.0.10
* Fixed bug affecting metrics with dimensions submitted when sample count was 0

### 0.0.9
* Adding ability to subfolder logs in S3

### 0.0.8
* Removing spacing in JSON stringify so logs in S3 are stored as one line

### 0.0.7
* Fixing a bug - accidentally published version where logs default timeout was 5 seconds

### 0.0.6
* Added the ability to copy logs to S3 using s3Bucket and s3Prefix options DO NOT USE - contains bug (see above)

### 0.0.5
* Added debug option to log to console all events

### 0.0.4
* Added multi-dimension support

### 0.0.3
* Fixing bug where instance id was hard-coded as a string rather than the actual instance ID

### 0.0.2
* Fixing bug where metrics were not sent due to a return statement before the cloudwatch call

### 0.0.1
* First release, beta

logs.js
var async = require('async');

var CloudWatchBuddyLogs = function(cloudwatchlogs, svc, s3, options){

    var api = {};

    var _logs = {};
    var _logsToSend;

    var _existingLogStreams = {};   // Holds known streams and their sequence tokens
    var _queuedSize = {};           // bytes to upload for each stream

    var _uploadInterval;

    var _logGroup = options.logGroup;
    var _timeout = (options.timeout && typeof options.timeout === 'number' && options.timeout >= 60 && options.timeout <= 1800) ? options.timeout : 120;
    var _maxSize = (options.maxSize && typeof options.maxSize === 'number' && options.maxSize < 1048576 && options.maxSize > 5000) ? options.maxSize : 200000;  // Default upload size of 200KB, AWS max of 1,048,576 bytes
    var _logFormat = (options.logFormat && typeof options.logFormat === 'string' && (options.logFormat === 'string' || options.logFormat === 'json')) ? options.logFormat : 'string';
    var _addTimestamp = (options.addTimestamp && typeof options.addTimestamp === 'boolean') ? options.addTimestamp : false;
    var _addInstanceId = (options.addInstanceId && typeof options.addInstanceId === 'boolean') ? options.addInstanceId : false;
    var _debug = (options.debug && typeof options.debug === 'boolean') ? options.debug : false;
    var _s3Subfolders = (options.s3Subfolders && typeof options.s3Subfolders === 'boolean') ? options.s3Subfolders : false;
    var _s3Bucket = (options.s3Bucket && typeof options.s3Bucket === 'string') ? options.s3Bucket : false;
    var _s3Prefix = (options.s3Prefix && typeof options.s3Prefix === 'string') ? options.s3Prefix : false;

    var _instanceId = 'unknown';

    // If _addInstanceId is set, then request the instance ID from AWS
    if (_addInstanceId) {
        svc.request('/latest/meta-data/instance-id', function(err, data){
            if (err && _debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : ERROR : Error retrieving instance ID from AWS : ' + err); }
            if (!err && data) {
                if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Instance ID retrieved from AWS : ' + data); }
                _instanceId = data; // set the instance ID in the background during initiation
            }
        });
    }

    var putLogData = function() {
        clearInterval(_uploadInterval);

        if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Put logs called'); }
        // Copy the log object so logs saved during the upload process aren't lost
        _logsToSend = JSON.parse(JSON.stringify(_logs));    // copy; don't reference
        _logs = {};                                         // Reset the logs instantly so none are lost
        
        // Reset the queued sizes
        for (key in _logsToSend) {
            _queuedSize[key] = 0;
        }

        async.eachSeries(Object.keys(_logsToSend), function(stream, callback){

            if (!_logsToSend[stream].length) {
                return callback();    // go to the next one
            }

            // If S3 option is selected, upload to S3 too
            if (_s3Bucket && _s3Prefix) {
                uploadLogsToS3(stream, _logsToSend[stream]); // Do this synchronously
            }

            checkIfLogStreamExistsAndCreateItIfItDoesNot(stream, function(err, data){
                if (err) {
                    if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : ERROR : Error checking if log stream exists : ' + err); }
                    callback(); // Don't callback an error - one log stream failure shouldn't affect the other streams
                } else {
                    // Stream now exists
                    if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Log stream exists : ' + stream); }
                    var params = {
                        logEvents: _logsToSend[stream],
                        logGroupName: _logGroup,
                        logStreamName: stream,
                        sequenceToken: _existingLogStreams[stream]
                    };
                    cloudwatchlogs.putLogEvents(params, function(err, data){
                        if (err) {
                            if (err.code === 'InvalidSequenceTokenException') {
                                if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : InvalidSequenceTokenException from AWS, retrying : ' + err); }
                                params.sequenceToken = err.message.substring(err.message.indexOf(':') + 2);
                                cloudwatchlogs.putLogEvents(params, function(err, data){
                                    if (err) {
                                        // Still having issues
                                        if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : ERROR : Error putting logs : ' + err); }
                                        callback(); // Don't callback an error for a single stream
                                    } else {
                                        _existingLogStreams[stream] = data.nextSequenceToken;   // Set this for next time
                                        callback(null, data);
                                    }
                                });
                            } else {
                                if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : ERROR : Error putting logs : ' + err); }
                                callback(); // An error for one stream shouldn't prevent others from being uploaded
                            }
                        } else {
                            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Successfully put logs for stream : ' + stream); }
                            _existingLogStreams[stream] = data.nextSequenceToken;   // Set this for next time
                            callback();
                        }
                    });
                }
            });
        }, function(err){
            if (err) {
                
            }
            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Finished putting logs. Resetting timer'); }
            setUploadInterval();    // Reset timer for next loop
        });
    };

    var uploadLogsToS3 = function(stream, logData) {
        if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Adding logs to S3 : ' + stream); }

        logFile = '';
        for (log in logData) {
            logFile += logData[log].message + '\n';
        }
        
        var timestamp = new Date();

        if (_s3Subfolders) {
            var key = _s3Prefix + '/' + stream + '/' + timestamp.getFullYear() + '/' + ('0' + (timestamp.getMonth()+1)).slice(-2) + '/' + ('0' + (timestamp.getDate())).slice(-2) + '/' + ('0' + (timestamp.getHours())).slice(-2) + '-' + ('0' + (timestamp.getMinutes())).slice(-2) + '-' + ('0' + (timestamp.getSeconds())).slice(-2) + '-' + timestamp.getMilliseconds() + '.log';
        } else {
            var key = _s3Prefix + '/' + stream + '/' + timestamp.getFullYear() + '-' + ('0' + (timestamp.getMonth()+1)).slice(-2) + '-' + ('0' + (timestamp.getDate())).slice(-2) + '-' + ('0' + (timestamp.getHours())).slice(-2) + '-' + ('0' + (timestamp.getMinutes())).slice(-2) + '-' + ('0' + (timestamp.getSeconds())).slice(-2) + '-' + timestamp.getMilliseconds() + '.log';
        }
        
        var params = {
            Bucket: _s3Bucket,
            Key: key,
            Body: logFile
        };

        s3.putObject(params, function(err, data){
            if (err && _debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : ERROR : S3 Error uploading logs : ' + stream + ' : ' + err); }
            else if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Logs uploaded to S3 successfully : ' + stream); }
        });
    };

    var checkIfLogStreamExistsAndCreateItIfItDoesNot = function(stream, callback) {
        if (_existingLogStreams[stream]) {
            callback();
        } else {
            var params = {
                logGroupName: _logGroup,
                logStreamNamePrefix: stream,
                limit: 1
            };
            cloudwatchlogs.describeLogStreams(params, function(err, data){
                if (!err && data.logStreams.length > 0) {
                    // The stream already exists, so add it to our known array and continue
                    // TODO: make sure the whole name matches, not just the prefix
                    _existingLogStreams[stream] = null; // Will eventually hold the sequence token
                    callback();
                } else {
                    // Create the stream
                    var params = {
                        logGroupName: _logGroup,
                        logStreamName: stream
                    };
                    cloudwatchlogs.createLogStream(params, callback);
                }
            });
        }
    };


    var setUploadInterval = function() {
        _uploadInterval = setInterval(function(){
            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Timer expired, calling put logs'); }
            putLogData();
        }, _timeout * 1000);
    };

    setUploadInterval();    // Call it first to start

    // Public functions

    api.log = function(stream, msg) {
        if (_logs[stream] === undefined) {
            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Adding new local log stream : ' + stream); }
            _logs[stream] = [];
        }

        if (_logFormat === 'string') {
            if (typeof msg === 'object') {
                msg = JSON.stringify(msg);
            }
            
            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Adding log string to local stream : ' + stream); }
            
            _logs[stream].push({
                timestamp: new Date().getTime(),
                message: (_addTimestamp ? new Date + ' ' : '') + (_addInstanceId ? _instanceId + ' ' : '') + msg
            });
        } else {
            var logObj = {};
            if (_addTimestamp) { logObj['timestamp'] = new Date; }
            if (_addInstanceId) { logObj['instance_id'] = _instanceId; }
            logObj['message'] = msg;

            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Adding log JSON to local stream : ' + stream); }

            _logs[stream].push({
                timestamp: new Date().getTime(),
                message: JSON.stringify(logObj)    // AWS only accepts a string
            });
        }

        _queuedSize[stream] += (26 + JSON.stringify(_logs[stream]).length * 2);    //~2 bytes per character plus 26 bytes of overhead per log
        
        if (((_queuedSize[stream]) >= (_maxSize - 1000)) || _logs[stream].length > 9000) {   // Leave some room (AWS max is 10,000 logs)
            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyLogs : INFO : Size of log queue for stream ' + stream + ' ' + _queuedSize[stream] + ' bytes is greater than max size of ' + _maxSize + ' bytes'); }
            putLogData();
        }
    };

    return api;
}

module.exports = CloudWatchBuddyLogs;

index.js
var AWS = require('aws-sdk');

var metricsHelper = require(__dirname + '/metrics.js');
var logsHelper = require(__dirname + '/logs.js');

var CloudWatchBuddy = function(aws){
    // Ensure a region is passed
    if (!aws || !aws.region) {
        throw new Error('Valid AWS config with a region is required');
    }

    AWS.config.update(aws);

    return {
        metrics: function(options) {
            if (!options || !options.namespace) {
                throw new Error('Valid metrics config with namespace is required');
            }

            if (options.namespace.indexOf('AWS/') === 0) {
                throw new Error('Namespace cannot begin with "AWS/"');
            }

            var cloudwatch = new AWS.CloudWatch(aws);
            return new metricsHelper(cloudwatch, options);
        },
        logs: function(options) {
            if (!options || !options.logGroup || typeof options.logGroup !== 'string') {
                throw new Error('Valid logs config with log group is required');
            }

            if (options.logGroup.length > 512) {
                throw new Error('Log group name cannot be more than the AWS limit of 512 characters');
            }

            var cloudwatchlogs = new AWS.CloudWatchLogs(aws);
            var svc = new AWS.MetadataService(aws);
            var s3 = new AWS.S3(aws);

            return new logsHelper(cloudwatchlogs, svc, s3, options);
        }
    }
}

module.exports = CloudWatchBuddy;

metrics.js
var validMetricValues = [
    'Seconds',
    'Microseconds',
    'Milliseconds',
    'Bytes',
    'Kilobytes',
    'Megabytes',
    'Gigabytes',
    'Terabytes',
    'Bits',
    'Kilobits',
    'Megabits',
    'Gigabits',
    'Terabits',
    'Percent',
    'Count',
    'Bytes/Second',
    'Kilobytes/Second',
    'Megabytes/Second',
    'Gigabytes/Second',
    'Terabytes/Second',
    'Bits/Second',
    'Kilobits/Second',
    'Megabits/Second',
    'Gigabits/Second',
    'Terabits/Second',
    'Count/Second',
    'None'
];

var CloudWatchBuddyMetrics = function(cloudwatch, options){

    var api = {};

    var _increments = {};
    var _stats = {};
    var _statsWithDimensions = [];

    var _uploadInterval;

    var _namespace = options.namespace;
    var _timeout = (options.timeout && typeof options.timeout === 'number' && options.timeout >= 60 && options.timeout <= 1800) ? options.timeout : 120;
    var _debug = (options.debug && typeof options.debug === 'boolean') ? options.debug : false;
    //var _maxSize = (typeof options.maxSize === 'number' && options.maxSize < 40000) ? options.maxSize : 40000;  // Max upload size if 40KB // TODO: add this option

    var putMetricData = function() {
        clearInterval(_uploadInterval);

        if (_debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : INFO : Put metrics called'); }

        var params = {
            MetricData:[],
            Namespace: _namespace
        };

        // Add the increments

        if (Object.keys(_increments).length > 0) {
            for (key in _increments) {
                var val = _increments[key];

                params.MetricData.push({
                    MetricName: key,
                    Timestamp: new Date,
                    Unit: 'Count',
                    Value: val,
                });
                _increments[key] = 0;  // reset for next time
            }
        }

        // Add the stats
        if (Object.keys(_stats).length > 0) {
            for (key in _stats) {
                if (_stats[key].SampleCount === 0) {
                    // This allows CWB to always send something, even if it is a 0 value
                    _stats[key].SampleCount = 1;
                }
                var obj = _stats[key];
                
                var pushObj = {
                    MetricName: key,
                    Timestamp: new Date,
                    Unit: obj.Unit,
                    StatisticValues: {
                        Maximum: obj.Maximum,
                        Minimum: obj.Minimum,
                        SampleCount: obj.SampleCount,
                        Sum: obj.Sum
                    }
                };

                params.MetricData.push(pushObj);

                // Reset the key
                _stats[key] = {
                    Maximum: 0,
                    Minimum: 0,
                    SampleCount: 0,
                    Sum: 0,
                    Unit: obj.Unit
                };
            }
        }

        // Add stats with dimensions
        if (_statsWithDimensions.length > 0) {
            for (index in _statsWithDimensions) {
                if (_statsWithDimensions[index].SampleCount === 0) {
                    // This allows CWB to always send something, even if it is a 0 value
                    _statsWithDimensions[index].SampleCount = 1;
                }
                var obj = _statsWithDimensions[index];
                
                var pushObj = {
                    MetricName: obj.MetricName,
                    Timestamp: new Date,
                    Unit: obj.Unit,
                    StatisticValues: {
                        Maximum: obj.Maximum,
                        Minimum: obj.Minimum,
                        SampleCount: obj.SampleCount,
                        Sum: obj.Sum
                    },
                    Dimensions: obj.Dimensions
                };

                params.MetricData.push(pushObj);

                _statsWithDimensions[index] = {
                    MetricName: obj.MetricName,
                    Unit: obj.Unit,
                    Maximum: 0,
                    Minimum: 0,
                    SampleCount: 0,
                    Sum: 0,
                    Dimensions: obj.Dimensions
                };
            }
        }

        if (params.MetricData.length > 0) {
            // TODO: Check size before sending and split if needed
            cloudwatch.putMetricData(params, function(err, data){
                if (err && _debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : ERROR : Put metrics error : ' + err); }
                if (!err && _debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : INFO : Put metrics success'); }
                // TODO: if err, see if retryable
                setUploadInterval();
            });
        } else {
            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : INFO : No metrics to put'); }
            setUploadInterval();
        }
    }

    var setUploadInterval = function() {
        _uploadInterval = setInterval(function(){
            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : INFO : Timer expired, calling put metrics'); }
            putMetricData();
        }, _timeout * 1000);
    }

    setUploadInterval();    // Call it first to start

    // Public functions

    api.increment = function(key){
        if (!key || typeof key !== 'string' || key.length < 1 || key.length > 100) { return; }

        if (_increments[key] === undefined) {
            _increments[key] = 0;
        }

        if (_debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : INFO : Incrementing metric : ' + key); }
        _increments[key]++;
    };

    api.stat = function(key, value, unit, dimensions) {
        if (!unit || validMetricValues.indexOf(unit) === -1) { 
            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : ERROR : Error adding stat : Invalid unit : ' + unit); }
            return;
        } // Only accept valid AWS metrics

        if (_stats[key] === undefined && !dimensions) {
            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : INFO : Adding stats key : ' + key); }
            _stats[key] = {
                Unit: unit,
                Maximum: value,
                Minimum: value,
                SampleCount: 1,
                Sum: value
            };
        } else if (dimensions && typeof dimensions === 'object' && Object.keys(dimensions).length > 0) {
            var convertedDimensions = [];

            for (name in dimensions) {
                if (typeof name !== 'string' || typeof dimensions[name] === 'object') {
                    if (_debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : ERROR : Invalid dimensions : ' + JSON.stringify(dimensions[name])); }
                    continue;
                }   // Only allow string - string / integer mappings

                convertedDimensions.push({
                    Name: name,
                    Value: dimensions[name]
                });
            }

            // Handle different dimensions sets
            var dimensionsFound = false;
            if (_statsWithDimensions.length > 0) {
                for (index in _statsWithDimensions) {
                    // Loop through each dimensions stat
                    if (JSON.stringify(_statsWithDimensions[index].Dimensions) === JSON.stringify(convertedDimensions)) {
                        // Edit the existing dimensions set
                        _statsWithDimensions[index].Maximum = value > _statsWithDimensions[index].Maximum ? value : _statsWithDimensions[index].Maximum;
                        _statsWithDimensions[index].Minimum = value < _statsWithDimensions[index].Minimum ? value : _statsWithDimensions[index].Minimum;
                        _statsWithDimensions[index].SampleCount++;
                        _statsWithDimensions[index].Sum += value;
                        dimensionsFound = true;
                        break;  // Don't continue the for loop once detected
                    }
                }
            }

            if (!dimensionsFound) {
                if (_debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : INFO : Adding stat with dimensions : ' + key); }

                _statsWithDimensions.push({
                    MetricName: key,
                    Unit: unit,
                    Maximum: value,
                    Minimum: value,
                    SampleCount: 1,
                    Sum: value,
                    Dimensions: convertedDimensions
                });
            }
        } else {
            if (_debug) { console.log (new Date() + ' : CloudWatchBuddyMetrics : INFO : Updating existing stats : ' + key); }

            _stats[key].Maximum = value > _stats[key].Maximum ? value : _stats[key].Maximum;
            _stats[key].Minimum = (value < _stats[key].Minimum || _stats[key].Minimum === 0) ? value : _stats[key].Minimum;
            _stats[key].SampleCount++;
            _stats[key].Sum += value;
        }
    };

    return api;
}

module.exports = CloudWatchBuddyMetrics;

example.js
var CloudWatchBuddy = require(__dirname + '/index.js');

var awsOptions = {
	accessKeyId: '',
	secretAccessKey: '',
	sessionToken: '',
	region: 'us-east-1'
};

var cwbMetricsOptions = {
	namespace: 'test-data',
	timeout: 60
};

var cwbLogsOptions = {
	logGroup: 'test-data',
	timeout: 60,
	addInstanceId: false,
	addTimestamp: true,
	logFormat: 'json' //|| 'json'
};

var cwbMetrics = new CloudWatchBuddy(awsOptions).metrics(cwbMetricsOptions);
var cwbLogs = new CloudWatchBuddy(awsOptions).logs(cwbLogsOptions);

cwbMetrics.increment('pageviews');
cwbMetrics.increment('pageviews');
cwbMetrics.increment('pageviews');
cwbMetrics.increment('pageviews');
cwbMetrics.increment('pageviews');

cwbMetrics.stat('loadtime', 10, 'Milliseconds');
cwbMetrics.stat('loadtime', 15, 'Milliseconds');
cwbMetrics.stat('loadtime', 7, 'Milliseconds');
cwbMetrics.stat('loadtime', 100, 'Milliseconds');

cwbMetrics.stat('pagesize', 10, 'Megabytes');

cwbMetrics.stat('serverload', 10, 'Percent', {serverName:'web01.example.com',region:'us-east-1'});
cwbMetrics.stat('serverload', 10, 'Percent', {serverName:'web02.example.com',region:'us-east-1'});

cwbMetrics.stat('serverload', 20, 'Percent', {serverName:'web01.example.com',region:'us-east-1'});
cwbMetrics.stat('serverload', 21, 'Percent', {serverName:'web01.example.com',region:'us-east-1'});
cwbMetrics.stat('serverload', 24, 'Percent', {serverName:'web01.example.com',region:'us-east-1'});
cwbMetrics.stat('serverload', 26, 'Percent', {serverName:'web01.example.com',region:'us-east-1'});
cwbMetrics.stat('serverload', 21, 'Percent', {serverName:'web01.example.com',region:'us-east-1'});
cwbMetrics.stat('serverload', 29, 'Percent', {serverName:'web01.example.com',region:'us-east-1'});
cwbMetrics.stat('serverload', 24, 'Percent', {serverName:'web01.example.com',region:'us-east-1'});
cwbMetrics.stat('serverload', 30, 'Percent', {serverName:'web03.example.com',region:'us-east-1'});
cwbMetrics.stat('serverload', 30, 'Percent', {serverName:'web03.example.com',region:'us-east-1',serverId:'xyz'});

//cwbLogs.log('errors', 'Dual message');
//cwbLogs.log('signups', 'Some user');

var i=0;
setInterval(function(){
	i++;
	cwbMetrics.stat('pagesize', 10, 'Megabytes');
	//cwbLogs.log('errors', 'Dual message: ' + i);
	//cwbLogs.log('signups', 'Some user: ' + i);
},500);